<div style="border: 3px solid #f55376; border-radius: 16px; background: linear-gradient(to right, #ffe4ec, #e0f7fa); padding: 20px; font-family: 'Segoe UI', sans-serif;">

<h2 style="color: #d81b60;">🧹 Data Cleaning & 🛠 Feature Engineering for Bank</h2>

<p style="font-size: 16px;">This notebook demonstrates the data cleaning and feature engineering steps applied to <b>customer, account, and transaction</b> data to prepare for downstream analysis and credit risk scoring.</p>

<ul style="font-size: 16px; line-height: 1.6;">
  <li>🧼 <b>Cleaning</b> phone numbers, email domains, and names</li>
  <li>💳 <b>Calculating</b> credit features like <code>loan_to_balance_ratio</code></li>
  <li>🧮 <b>Generating</b> a synthetic <code>cibil_score</code></li>
  <li>🧾 <b>Labeling</b> customers into <code>credit risk bands</code></li>
</ul>

<p style="font-style: italic; font-size: 15px; margin-top: 10px;">
This notebook is part of a two-part EDA pipeline. It performs all necessary cleaning and feature engineering before final reporting and analysis are done in the main EDA notebook. 💼📊
</p>

</div>

## 🧭 Table of Contents

---

1. Customer Table: Raw Data Overview
2. Name Cleaning
3. Email Domain Extraction
4. Phone Number Cleaning
5. Address Parsing
6. Credit Features (loan ratio, CIBIL and level of indicators)

## 🔌⚙️Reaching out to Sql Alchemy:
⬇ Tooling up with seamless DB integration

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# Connect to MySQL
from sqlalchemy import create_engine
engine = create_engine("mysql+pymysql://root:nikitasql@localhost/github_db")

## 1.🙍‍♂️ Customer Table : A raw data Overview

In [32]:
# Load datasets
df_customer_raw = pd.read_sql("Select * FROM customer_raw",engine)
df_customer_raw.head()

customer_id             name  age  gender  \
0            1     Allison Hill   69   Other   
1            2      Noah Rhodes   32  Female   
2            3  Angie Henderson   78  Female   
3            4    Daniel Wagner   38    Male   
4            5  Cristian Santos   41  Female   

                                   email                phone  \
0                christopher86@gmail.com  +1-111-109-6179x817   
1                        cwood@gmail.com   904.633.7861x37157   
2                    omarshall@yahoo.com                        
3             karencontreras@barnett.biz           2589552963   
4  randallloretta@sheppard-mcdaniel.info                -7509   

                                             address  branch_id cleaned_phone  
0                722 Erica Trace, Kingfort, DE 70686          1          None  
1  07681 Ian Heights Apt. 855, Camachohaven, AL 2...          7          None  
2  23622 Peter Camp Suite 968, Melanieview, TN 62387          8          None  
3  516 John Vista Suite 215, Lake Darrellchester,...          3          None  
4          0335 Tiffany Via, Patriciaburgh, IL 92620          4          None

## 🔍 Identified Data Quality Issues :

## 📌 Follow-Up Actions in Code!

## ------------------------------------------------------------------------------------------------

### ✅ Checking the duplicate/unique names from customer_raw table.

In [33]:
query = '''
        select Count(*),count(distinct name) 
        from customers_raw  
        ;
        '''
name = pd.read_sql(query,engine)
name

Count(*)  count(distinct name)
0      1102                   995

### Using row_number() to remove duplicate values using CTE:

In [ ]:
from sqlalchemy import text

with engine.begin() as conn:
    conn.execute(text("""
        CREATE TABLE aligned_customers_ws AS
        SELECT *
        FROM (
            SELECT *, 
                   ROW_NUMBER() OVER (PARTITION BY customer_id ORDER BY customer_id) AS rn
            FROM customers_raw
        ) AS t
        WHERE t.rn = 1;
    """))

cleaned_name_rn = pd.read_sql(query,engine)
cleaned_name_rn

### ❓Why duplication is not necessarily a dirty data

- Remaining 5 names are duplicate but distinct with different respective details.

In [24]:
query = '''
            select count(*),count(distinct customer_id),count(distinct name) from aligned_customers_ws
        '''

dupe = pd.read_sql(query,engine)
dupe

count(*)  count(distinct customer_id)  count(distinct name)
0      1000                         1000                   995

In [40]:
query = '''
            SELECT *
            FROM aligned_customers_ws 
            WHERE name in(select name from aligned_customers_ws GROUP BY name HAVING COUNT(*) > 1)
            order by name
            ;
        '''

cust_c = pd.read_sql(query,engine)
cust_c

customer_id              name  age  gender  \
0  CUST000238       David Davis   44  Female   
1  CUST000540       David Davis   64  Female   
2  CUST000620  Elizabeth Mendez   25    Male   
3  CUST000788  Elizabeth Mendez   68    Male   
4  CUST000099     Jessica Smith   20    Male   
5  CUST000764     Jessica Smith   51   Other   
6  CUST000225     Matthew Moore   25    Male   
7  CUST000276     Matthew Moore   44  Female   
8  CUST000626    Michael Miller   84  Female   
9  CUST000868    Michael Miller   18  Female   

                                  email                phone  \
0                      no email updated     763.369.1128x981   
1              selenarobinson@gmail.com         531-234-5506   
2                colonbobby@jackson.com   (022)132-1767x2019   
3   williamsimmons@collins-griffith.com         695.528.6209   
4                   jmckinney@gmail.com           1935889536   
5                   rtaylor@jackson.org                        
6               cynthianguyen@bauer.com         499.601.0548   
7                 michael43@hotmail.com  (889)362-8969x64546   
8  lauriegutierrez@villanueva-crane.net  +1-529-701-3712x436   
9                    dustin34@gmail.com                        

                                             address  branch_id  \
0  1622 Bennett Stravenue Suite 718, Michaelport,...          1   
1                   Unit 9908 Box 6723, DPO AP 29081          2   
2                   PSC 1862, Box 4475, APO AA 25149          5   
3  2951 Lopez Lake Suite 768, West Rebeccashire, ...          9   
4                  0918 Hicks Key, Amyland, WA 13010          7   
5     587 Cruz Green Suite 715, Meaganbury, NV 33628          9   
6             8912 Jeffrey Shore, Kingbury, VA 53257         10   
7  3866 Shelton Garden Apt. 031, Schwartzchester,...          5   
8  72876 Jennifer Drive Apt. 877, Port Joseph, WA...          8   
9  164 Benson Ports Apt. 532, New Diamondville, I...          4   

                                          street      cleaned_phone  rn  \
0  1622 Bennett Stravenue Suite 718, Michaelport    7633691128 x981   1   
1               Unit 9908 Box 6723, DPO AP 29081         5312345506   1   
2               PSC 1862, Box 4475, APO AA 25149   0221321767 x2019   1   
3   2951 Lopez Lake Suite 768, West Rebeccashire         6955286209   1   
4                        0918 Hicks Key, Amyland         1935889536   1   
5           587 Cruz Green Suite 715, Meaganbury               None   1   
6                   8912 Jeffrey Shore, Kingbury         4996010548   1   
7  3866 Shelton Garden Apt. 031, Schwartzchester  8893628969 x64546   1   
8     72876 Jennifer Drive Apt. 877, Port Joseph  +15297013712 x436   1   
9    164 Benson Ports Apt. 532, New Diamondville               None   1   

       cleaned_name          email_domain domain_status domain_type  \
0       David Davis      no email updated       Invalid     no type   
1       David Davis             gmail.com         Valid    personal   
2  Elizabeth Mendez           jackson.com         Valid    business   
3  Elizabeth Mendez  collins-griffith.com       Invalid    business   
4     Jessica Smith             gmail.com         Valid    personal   
5     Jessica Smith           jackson.org         Valid    business   
6     Matthew Moore             bauer.com         Valid    business   
7     Matthew Moore           hotmail.com         Valid    personal   
8    Michael Miller  villanueva-crane.net       Invalid    business   
9    Michael Miller             gmail.com         Valid    personal   

  state_code                  state zipcode      country loan_to_balance_ratio  
0        VA                Virginia   59417           US                  None  
1        AP    Armed Forces Pacific   29081  US military                  None  
2        AA   Armed Forces Americas   25149  US military                  None  
3        IA                    Iowa    7531           US                  None  
4        WA      

## 2.🙍‍♂️Cleaning Names using Regexp to maintain Standardisation:

In [ ]:
#Altering table to add a column : cleaned_name

with engine.begin() as conn:
    conn.execute(text("""
        alter table aligned_customers_ws add column cleaned_name varchar(54);
    """))
# cleaned_name= pd.read_sql(query,engine)
# cleaned_name

In [43]:
#Select statement to check the regexp query:

query = '''
        select name ,
        trim(trim('.' from regexp_replace(cleaned_name,'(?i)\\b(MD|Dr\\.?|Dr.\\?|DVM|Mr\\.?|Mrs\\.?|Ms\\.?|Miss|Jr\\.?|Sr\\.?|II|III||PHD|DDS|DMD|MBA|Esq\\.?)\\b',''))) as cleaned_name
        from aligned_customers_ws 
        where name<>trim(trim('.' from regexp_replace(cleaned_name,'(?i)\\b(MD|Dr\\.?|Dr.\\?|DVM|Mr\\.?|Mrs\\.?|Ms\\.?|Miss|Jr\\.?|Sr\\.?|II|III||PHD|DDS|DMD|MBA|Esq\\.?)\\b','')))
        ;
        '''
col_name = pd.read_sql(query,engine)
col_name.head(10)

name      cleaned_name
0  Margaret Hawkins DDS  Margaret Hawkins
1     David Douglas Jr.     David Douglas
2  Dr. Hannah Patterson  Hannah Patterson
3        Ricky Davis II       Ricky Davis
4    Dr. William Warren    William Warren
5      Bailey Duran DDS      Bailey Duran
6  Elizabeth Oliver DDS  Elizabeth Oliver
7     Brandon Bailey MD    Brandon Bailey
8       Susan Murray MD      Susan Murray
9    Mrs. Kristen Reyes     Kristen Reyes

In [ ]:
#Updating query for the above trial:

from sqlalchemy import text

with engine.begin() as conn:
    conn.execute(text("""
        UPDATE aligned_customers_ws 
        SET cleaned_name = REGEXP_REPLACE(
            name,
            '(?i)\\b(MD|Dr\\.?|Mr\\.?|Mrs\\.?|Ms\\.?|Miss|Jr\\.?|Sr\\.?|PHD|DDS|DMD|MBA|Esq\\.?)\\b',
            ''
        )
        WHERE cleaned_name IS NULL;
    """))



with engine.begin() as conn:
    conn.execute(text("""
        update aligned_customers_ws 
        set cleaned_name = trim(trim('.' from regexp_replace(cleaned_name,'(?i)\\b(MD|Dr\\.?|Dr.\\?|DVM|Mr\\.?|Mrs\\.?|Ms\\.?|Miss|Jr\\.?|Sr\\.?|II|III||PHD|DDS|DMD|MBA|Esq\\.?)\\b','')))
        where cleaned_name <>trim(trim('.' from regexp_replace(cleaned_name,'(?i)\\b(MD|Dr\\.?|Dr.\\?|DVM|Mr\\.?|Mrs\\.?|Ms\\.?|Miss|Jr\\.?|Sr\\.?|II|III||PHD|DDS|DMD|MBA|Esq\\.?)\\b','')))
        and cleaned_name is not null;
        """
        ))

In [44]:
#Select statement to check updates:

query = '''
        select name ,cleaned_name
        from aligned_customers_ws 
        where cleaned_name is not null and (name like '%%Jr%%' OR name like '%%Dr%%')
        ;
        '''
col_name = pd.read_sql(query,engine)
col_name

name         cleaned_name
0         Anthony Rodriguez    Anthony Rodriguez
1          Cassandra Gaines     Cassandra Gaines
2           Brian Rodriguez      Brian Rodriguez
3            Sandra Aguilar       Sandra Aguilar
4         Richard Rodriguez    Richard Rodriguez
5         David Douglas Jr.        David Douglas
6        Patricia Rodriguez   Patricia Rodriguez
7      Dr. Hannah Patterson     Hannah Patterson
8        Dr. William Warren       William Warren
9           Andrew Reynolds      Andrew Reynolds
10             Andrew Avila         Andrew Avila
11             Donna Landry         Donna Landry
12          Dr. Paul Morgan          Paul Morgan
13           Andrea Hubbard       Andrea Hubbard
14          Kayla Rodriguez      Kayla Rodriguez
15       Theodore Jones Jr.       Theodore Jones
16       Courtney Rodriguez   Courtney Rodriguez
17        Mr. Andrew Foster        Andrew Foster
18             Sandra Drake         Sandra Drake
19         Kelsey Rodriguez     Kelsey Rodriguez
20    Dr. Kelly Hammond DVM        Kelly Hammond
21        Dr. Michael Evans        Michael Evans
22  Mr. Alexander Landry MD     Alexander Landry
23          Keith Rodriguez      Keith Rodriguez
24          Molly Rodriguez      Molly Rodriguez
25            Andrew Harper        Andrew Harper
26              Ryan Landry          Ryan Landry
27      Alexandra Dominguez  Alexandra Dominguez
28            Sandra Becker        Sandra Becker
29      Dr. Andrea Marshall      Andrea Marshall
30            Sandra Rivera        Sandra Rivera
31          Kendra Wang DVM          Kendra Wang
32           Eric Henry Jr.           Eric Henry
33          David Rodriguez      David Rodriguez

### 🔍 Post-Name Cleaning Validation


In [45]:
query = '''
            select * from aligned_customers_ws where cleaned_name is null;
        '''
name = pd.read_sql(query, engine)
name

Empty DataFrame
Columns: [customer_id, name, age, gender, email, phone, address, branch_id, street, cleaned_phone, rn, cleaned_name, email_domain, domain_status, domain_type, state_code, state, zipcode, country, loan_to_balance_ratio]
Index: []

All 1000 records in the `name` column were successfully cleaned.
- `cleaned_name` column has **no NULL values**
- All known prefixes and suffixes were removed using regex.


## 3.📧 Email Domain Extraction to recognize Status & Types :

In [ ]:
#Altering the table to add a column email_domain:

with engine.begin() as conn:
    conn.execute(text('''
                        ALter table aligned_customers_ws ADD COLUMN email_domain varchar(80);
                        '''
                ))

In [26]:
#Updating the email_domain column by extracting from email column:

with engine.connect() as conn:
    conn.execute(text("""
        UPDATE aligned_customers_ws
        SET email_domain = SUBSTRING_INDEX(email, '@', -1);
    """))

query = '''
            select email, email_domain 
            from aligned_customers_ws
            ;
        '''
domain = pd.read_sql(query, engine)
domain

email            email_domain
0                  christopher86@gmail.com               gmail.com
1                          cwood@gmail.com               gmail.com
2                      omarshall@yahoo.com               yahoo.com
3               karencontreras@barnett.biz             barnett.biz
4    randallloretta@sheppard-mcdaniel.info  sheppard-mcdaniel.info
..                                     ...                     ...
995            woodmichael@mosley-shaw.com         mosley-shaw.com
996                   wgutierrez@yahoo.com               yahoo.com
997                     daniel27@yahoo.com               yahoo.com
998                 angelamartin@yahoo.com               yahoo.com
999                       no email updated        no email updated

[1000 rows x 2 columns]

### (i) ✅ Verifying the email domains through DNS(domain name system)

- DNS translates domain names into IP address like 198.168.1. to build connections with other machines.

### (ii) What is Domain Lookup Doing ❓

For an email like 'allison.com', domain lookup would:
  1) Extract the domain: gmail.com
  2) Query the DNS system to check:
     - Is gmail.com a valid domain?
     - Does it have MX records (i.e. mail servers configured)?
  3) If it has MX records → ✅ Valid email domain
  4) If not → ❌ Invalid domain (likely not able to receive mail)

In [48]:
# ! pip install dnspython

import dns.resolver 

def is_valid_domain(domain):
    try:
        dns.resolver.resolve(domain, 'MX') 
        return 'Valid'
    except:
        return 'Invalid'
#Mail Exchange is one of the record types, that tells email where to be delivered.


#testing email_domain
is_valid_domain('allison.com')

'Valid'

In [ ]:
import pandas as pd

query = '''
            select * from aligned_customers_ws;
        '''
df = pd.read_sql(query,engine)
df

In [257]:
#Appling Validation of domains in domain_status column using a function

df['domain_status'] = df['email_domain'].apply(is_valid_domain)
df['domain_status']

0        Valid
1        Valid
2        Valid
3      Invalid
4      Invalid
        ...   
995    Invalid
996      Valid
997      Valid
998      Valid
999    Invalid
Name: domain_status, Length: 1000, dtype: object

In [49]:
# Check a few rows to see the results

print(df[['email', 'email_domain', 'domain_status']].tail(10))

                           email      email_domain domain_status
990         sharpbeth@morris.com        morris.com         Valid
991     peggypeck@lin-wilson.biz    lin-wilson.biz       Invalid
992       mcneilsherry@gmail.com         gmail.com         Valid
993       johnsonjulie@gmail.com         gmail.com         Valid
994    carolynrobinson@gmail.com         gmail.com         Valid
995  woodmichael@mosley-shaw.com   mosley-shaw.com       Invalid
996         wgutierrez@yahoo.com         yahoo.com         Valid
997           daniel27@yahoo.com         yahoo.com         Valid
998       angelamartin@yahoo.com         yahoo.com         Valid
999             no email updated  no email updated       Invalid


## Updating the email_domains and status in sql table:
As the above pandas updates exist only in Python session — it does not update the database unless it's done:

In [ ]:
# Push the updates to a temporary table

df[['customer_id', 'domain_status']].to_sql(
    name='temp_domain_status',
    con=engine,
    index=False,         # Doesn't need DF's index as a new column
    if_exists='replace'  # Replace any existing temp data
)
df

In [50]:
#Testing Temp Table:

query = '''
        SELECT * FROM temp_domain_status LIMIT 10;
        '''
test = pd.read_sql(query , engine)
test

customer_id domain_status
0  CUST000001         Valid
1  CUST000002         Valid
2  CUST000003         Valid
3  CUST000004       Invalid
4  CUST000005       Invalid
5  CUST000006         Valid
6  CUST000007         Valid
7  CUST000008       Invalid
8  CUST000009         Valid
9  CUST000010         Valid

In [260]:
#Updating the columns by joining Temp Table:

from sqlalchemy import text

with engine.begin() as conn:
    conn.execute(text("""
        UPDATE aligned_customers_ws 
        JOIN temp_domain_status ON aligned_customers_ws.customer_id = temp_domain_status.customer_id
        SET aligned_customers_ws.domain_status = temp_domain_status.domain_status;
    """))

In [51]:
#Checking the updated columns with relevant email columns using `for` loop:

with engine.begin() as conn:
    result = conn.execute(text("""
        SELECT customer_id, email_domain, domain_status 
        FROM aligned_customers_ws 
        LIMIT 10;
    """))
    for row in result:
        print(row)

('CUST000001', 'gmail.com', 'Valid')
('CUST000002', 'gmail.com', 'Valid')
('CUST000003', 'yahoo.com', 'Valid')
('CUST000004', 'barnett.biz', 'Invalid')
('CUST000005', 'sheppard-mcdaniel.info', 'Invalid')
('CUST000006', 'johnson.org', 'Valid')
('CUST000007', 'stanton.com', 'Valid')
('CUST000008', 'harris-carson.com', 'Invalid')
('CUST000009', 'yahoo.com', 'Valid')
('CUST000010', 'gmail.com', 'Valid')


In [28]:
#Count of email and derived columns:

query = '''
        SELECT count(email), count(email_domain) , count(domain_status)
        FROM aligned_customers_ws; 
        '''
em_cust = pd.read_sql(query,engine)
em_cust

count(email)  count(email_domain)  count(domain_status)
0          1000                 1000                  1000

In [61]:
#Updating the domain status with `no domain` and email_domain with `no email updated` for missing emails:


# with engine.begin() as conn:
#     conn.execute(text(
#                         '''
#                         update aligned_customers_ws 
#                         set domain_status = 'no domain'
#                         where email = 'no email updated' and domain_status is not NULL
#                         '''
#     ))


query ='''
        select customer_id,email_domain,domain_status
        from aligned_customers_ws 
        where email = 'no email updated'
        limit 10;
        '''
df = pd.read_sql(query,engine)
df  

customer_id      email_domain domain_status
0  CUST000063  no email updated     no domain
1  CUST000065  no email updated     no domain
2  CUST000067  no email updated     no domain
3  CUST000071  no email updated     no domain
4  CUST000082  no email updated     no domain
5  CUST000117  no email updated     no domain
6  CUST000128  no email updated     no domain
7  CUST000156  no email updated     no domain
8  CUST000176  no email updated     no domain
9  CUST000177  no email updated     no domain

## Testing Different Domains and Status:

In [29]:
#For Business Domains:

query = '''
        select email,email_domain,domain_status,domain_type
        from aligned_customers_ws
        WHERE domain_type = 'business' 
        ;
        '''

bs_domain = pd.read_sql(query,engine)
bs_domain

email            email_domain  \
0               karencontreras@barnett.biz             barnett.biz   
1    randallloretta@sheppard-mcdaniel.info  sheppard-mcdaniel.info   
2                       zevans@johnson.org             johnson.org   
3                   averyshane@stanton.com             stanton.com   
4                rmiller@harris-carson.com       harris-carson.com   
..                                     ...                     ...   
463      davisrachel@richmond-phillips.com   richmond-phillips.com   
464                       sgomez@price.org               price.org   
465                   sharpbeth@morris.com              morris.com   
466               peggypeck@lin-wilson.biz          lin-wilson.biz   
467            woodmichael@mosley-shaw.com         mosley-shaw.com   

    domain_status domain_type  
0         Invalid    business  
1         Invalid    business  
2           Valid    business  
3           Valid    business  
4         Invalid    business  
..            ...         ...  
463       Invalid    business  
464         Valid    business  
465         Valid    business  
466       Invalid    business  
467       Invalid    business  

[468 rows x 4 columns]

In [31]:
# For Personal Domains:

query = '''
        select email,email_domain,domain_status,domain_type
        from aligned_customers_ws
        where domain_type = 'personal'
        ;
        '''
ps_dom = pd.read_sql(query,engine)
ps_dom

email email_domain domain_status domain_type
0      christopher86@gmail.com    gmail.com         Valid    personal
1              cwood@gmail.com    gmail.com         Valid    personal
2          omarshall@yahoo.com    yahoo.com         Valid    personal
3            renee11@yahoo.com    yahoo.com         Valid    personal
4          melissa87@gmail.com    gmail.com         Valid    personal
..                         ...          ...           ...         ...
480     johnsonjulie@gmail.com    gmail.com         Valid    personal
481  carolynrobinson@gmail.com    gmail.com         Valid    personal
482       wgutierrez@yahoo.com    yahoo.com         Valid    personal
483         daniel27@yahoo.com    yahoo.com         Valid    personal
484     angelamartin@yahoo.com    yahoo.com         Valid    personal

[485 rows x 4 columns]

In [32]:
# For missing emails:
query ='''
        select customer_id,email_domain,domain_status
        from aligned_customers_ws 
        where email = 'no email updated'
        limit 10;
        '''
df = pd.read_sql(query,engine)
df.head(10)

customer_id      email_domain domain_status
0  CUST000063  no email updated     no domain
1  CUST000065  no email updated     no domain
2  CUST000067  no email updated     no domain
3  CUST000071  no email updated     no domain
4  CUST000082  no email updated     no domain
5  CUST000117  no email updated     no domain
6  CUST000128  no email updated     no domain
7  CUST000156  no email updated     no domain
8  CUST000176  no email updated     no domain
9  CUST000177  no email updated     no domain

## Post Email Cleaning Validation:

- Total records with email: 1000  
- Personal domains (gmail/yahoo): 485 — all valid  
- Business domains: 468  
  - Valid: 225  
  - Invalid: 243 
- Missing/Placeholder emails: 47 → replaced with 'no email updated' to retain record integrity during joins

## 4.📞 Cleaning Phone For Standardization

In [ ]:
#Adding a column to insert cleaned version of phone numbers:

with engine.begin() as conn:
    conn.execute(text(
        '''
        alter table aligned_customers_ws add column cleaned_phone VARCHAR(54);
        '''
    ))

In [6]:
#Updating cleaned_phone column using regex queries:

from sqlalchemy import text
with engine.begin() as conn:
    txt = conn.execute(text(
                # '''
                # update aligned_customers_ws 
                # set cleaned_phone = case when substring_index(regexp_replace(phone,'[(+\\.)-]',''),'x',1) like '001%%' then 
                # regexp_replace(substring_index(regexp_replace(phone,'[(+\\.)-]',''),'x',1),'001','+1 ') 
                # else null end
                # where length(phone)>10
                # and phone like '001%%'
                # and cleaned_phone is null
                # ;
                # '''

                # '''
                # update aligned_customers_ws 
                # set cleaned_phone = concat(cleaned_phone,' x',substring_index(phone,'x',-1))
                # where cleaned_phone is not null
                # ;
                # '''

                # '''
                # UPDATE aligned_customers_ws as t1
                # JOIN cleaned_phone_numbers as t2 ON t1.phone = t2.phone
                # SET t1.cleaned_phone = case WHEN LENGTH(t2.phone_numbers) < 10 THEN NULL   
                # ELSE CONCAT(SUBSTRING_INDEX(t2.phone_numbers, 'x', 1), ' x',SUBSTRING_INDEX(t2.phone_numbers, 'x', -1))END
                # WHERE t1.cleaned_phone IS NULL
                # AND t2.phone_numbers LIKE '%x%'
                #  ;
                # '''

               '''    
                UPDATE aligned_customers_ws as t1
                JOIN cleaned_phone_numbers as t2 ON t1.phone = t2.phone
                SET t1.cleaned_phone = 'No phone length'
                where t1.cleaned_phone is null and length(Phone_numbers)<10 and Phone_numbers is not null
                ;
              '''
    ))
txt
result_df = pd.read_sql("""
    SELECT phone, cleaned_phone
    FROM aligned_customers_ws
    WHERE cleaned_phone = 'No phone length'
""", engine)
result_df.head()

phone    cleaned_phone
0      -7509  No phone length
1  215632135  No phone length
2      -4158  No phone length
3      -5483  No phone length
4      -9246  No phone length

## 🔢 Phone-Digit Variation 📞

1. Starts with +1 → International Format for USA/Canada:
+1 = Country Code for 🇺🇸 USA and 🇨🇦 Canada (North American Numbering Plan)
Often seen in:
    - Business contacts
    - International calls
    - CRM/ERP systems
------------------------------------------------------------------------------------------------------------------
2. Starts with 0 → Local/National Format (No Country Code):
    - The 0 is a trunk prefix — it’s used for national dialing only
    - Not valid for international calling unless country code is added
------------------------------------------------------------------------------------------------------------------
3. Includes x or ext → Internal Routing in Organizations:
    - `x` or `ext` stands for extension.
    - Used in `corporate` or `institutional phone systems` to reach specific people or departments.
    - Caller usually hears an IVR: “If you know your party’s extension, please dial it now.”


## 🔍 Post-Phone Cleaning Validation

In [7]:
query = '''
        SELECT COUNT(cleaned_phone),count(DISTINCT cleaned_phone)
        FROM aligned_customers_ws 
        WHERE cleaned_phone IS NOT NULL;
        '''
phone = pd.read_sql(query,engine)
phone

COUNT(cleaned_phone)  count(DISTINCT cleaned_phone)
0                   948                            902

- **Total Count of Phone: 1000**
  
- **Total Count of Not NULL Phone : 948**

  **Reason:** Out of 1000, **52 rows had blank or missing entries** in the original `phone` column, making them unfixable via standard parsing. Hence, it was   deliberately retained as **NULL** to preserve row integrity while signaling incomplete data.

- **Total Count of Unique Cleaned_Phone : 902**

  📱 cleaned_phone had 47 missing entries labeled 'No phone length', resulting in 901 valid distinct values out of 948 total.



## 5. 🏠 Parsing Address Fields: Street from Address

In [8]:
# Adding columns to split the address into street,state_code,state and zipcode

with engine.begin() as conn:
    txt = conn.execute(text(
                    '''
                    -- SELECT COUNT(distinct address) 
                    # alter table aligned_customers_ws add column street varchar(60);
                    # alter table aligned_customers_ws add column state_code varchar(60);
                    # alter table aligned_customers_ws add column state varchar(60);
                    # alter table aligned_customers_ws add column zipcode varchar(60);
                    # alter table aligned_customers_ws add column country varchar(60);
                    '''
                    ))
txt
query = '''
        select address, street, state_code, state, zipcode
        from aligned_customers_ws
        ;
        '''
add_st = pd.read_sql(query,engine)
add_st

address  \
0                  722 Erica Trace, Kingfort, DE 70686   
1    07681 Ian Heights Apt. 855, Camachohaven, AL 2...   
2    23622 Peter Camp Suite 968, Melanieview, TN 62387   
3    516 John Vista Suite 215, Lake Darrellchester,...   
4            0335 Tiffany Via, Patriciaburgh, IL 92620   
..                                                 ...   
995  51547 Graham Crescent Apt. 708, Smithfort, ND ...   
996                   PSC 4009, Box 3808, APO AP 26991   
997        32160 Nicholas Rapid, East Joseph, ND 31097   
998  76252 Nicole Corners Apt. 116, Elliottport, NC...   
999  0116 Wagner Creek Suite 786, New Adrian, CT 44160   

                                            street state_code  \
0                        722 Erica Trace, Kingfort        DE    
1         07681 Ian Heights Apt. 855, Camachohaven        AL    
2          23622 Peter Camp Suite 968, Melanieview        TN    
3    516 John Vista Suite 215, Lake Darrellchester        ME    
4                  0335 Tiffany Via, Patriciaburgh        IL    
..                                             ...        ...   
995      51547 Graham Crescent Apt. 708, Smithfort        ND    
996               PSC 4009, Box 3808, APO AP 26991        AP    
997              32160 Nicholas Rapid, East Joseph        ND    
998     76252 Nicole Corners Apt. 116, Elliottport        NC    
999        0116 Wagner Creek Suite 786, New Adrian        CT    

                    state zipcode  
0                Delaware   70686  
1                 Alabama   22746  
2               Tennessee   62387  
3                   Maine   19531  
4                Illinois   92620  
..                    ...     ...  
995          North Dakota   24929  
996  Armed Forces Pacific   26991  
997          North Dakota   31097  
998        North Carolina   98164  
999           Connecticut   44160  

[1000 rows x 5 columns]

### For Street :

In [9]:
with engine.begin() as conn:
    conn.execute(text(
        '''
            # select address,substring_index(address,',',2)
            # from aligned_customers_ws 
            # where address  like 'psc%'
            # ;
            
            # select address ,substring_index(address,',',3), street
            # from aligned_customers_ws where address like 'psc%';            
            
            update aligned_customers_ws
            set street = substring_index(address,',',3)
            where address not like 'psc%' and street is null  
            ;
        '''
    ))
    
query = '''
            select address , street from aligned_customers_ws
            where street is not null and address is not null;
        '''
query = pd.read_sql(query,engine)
query

address  \
0                  722 Erica Trace, Kingfort, DE 70686   
1    07681 Ian Heights Apt. 855, Camachohaven, AL 2...   
2    23622 Peter Camp Suite 968, Melanieview, TN 62387   
3    516 John Vista Suite 215, Lake Darrellchester,...   
4            0335 Tiffany Via, Patriciaburgh, IL 92620   
..                                                 ...   
995  51547 Graham Crescent Apt. 708, Smithfort, ND ...   
996                   PSC 4009, Box 3808, APO AP 26991   
997        32160 Nicholas Rapid, East Joseph, ND 31097   
998  76252 Nicole Corners Apt. 116, Elliottport, NC...   
999  0116 Wagner Creek Suite 786, New Adrian, CT 44160   

                                            street  
0                        722 Erica Trace, Kingfort  
1         07681 Ian Heights Apt. 855, Camachohaven  
2          23622 Peter Camp Suite 968, Melanieview  
3    516 John Vista Suite 215, Lake Darrellchester  
4                  0335 Tiffany Via, Patriciaburgh  
..                                             ...  
995      51547 Graham Crescent Apt. 708, Smithfort  
996               PSC 4009, Box 3808, APO AP 26991  
997              32160 Nicholas Rapid, East Joseph  
998     76252 Nicole Corners Apt. 116, Elliottport  
999        0116 Wagner Creek Suite 786, New Adrian  

[1000 rows x 2 columns]

### For state_code:

In [10]:
with engine.begin() as conn:
    conn.execute(text(
        '''
            #         select address, 
            # regexp_replace(substring_index(address,' ',-2),'[0-9]','')  
            # from aligned_customers_ws 
            # where address not like 'PSC%'
            # and address not like 'US%'
            # and address not like 'Unit%'
            # ; 
            
            update aligned_customers_ws
            set state_code = regexp_replace(substring_index(address,' ',-2),'[0-9]','')  
            where address not like 'PSC%'
            and address not like 'US%'
            and address not like 'Unit%'
            and state_code is null
            ;
        '''
    ))
query = '''
            select address, state, state_code
            from aligned_customers_ws
        '''
state_add = pd.read_sql(query,engine)
state_add

address                 state  \
0                  722 Erica Trace, Kingfort, DE 70686              Delaware   
1    07681 Ian Heights Apt. 855, Camachohaven, AL 2...               Alabama   
2    23622 Peter Camp Suite 968, Melanieview, TN 62387             Tennessee   
3    516 John Vista Suite 215, Lake Darrellchester,...                 Maine   
4            0335 Tiffany Via, Patriciaburgh, IL 92620              Illinois   
..                                                 ...                   ...   
995  51547 Graham Crescent Apt. 708, Smithfort, ND ...          North Dakota   
996                   PSC 4009, Box 3808, APO AP 26991  Armed Forces Pacific   
997        32160 Nicholas Rapid, East Joseph, ND 31097          North Dakota   
998  76252 Nicole Corners Apt. 116, Elliottport, NC...        North Carolina   
999  0116 Wagner Creek Suite 786, New Adrian, CT 44160           Connecticut   

    state_code  
0          DE   
1          AL   
2          TN   
3          ME   
4          IL   
..         ...  
995        ND   
996        AP   
997        ND   
998        NC   
999        CT   

[1000 rows x 3 columns]

### For Country:

In [11]:
from sqlalchemy import text

with engine.begin() as conn:
        conn.execute(text(
            '''
            # update aligned_customers_ws 
            # set country = 'US military' 
            # where address like 'PSC%'
            # OR address like 'US%'
            # OR address like 'Unit%'
            # and state_code is not null 
            # and country is NULL
            # ; 
            
            update aligned_customers_ws 
            set country = 'US'
            where country is null;
            '''
))
# Military mail system, which is operated by the U.S. Postal Service in cooperation with the
# U.S. Department of Defense.
query = '''
        select address ,country
        from aligned_customers_ws;
        '''
country = pd.read_sql(query,engine)
country

address      country
0                  722 Erica Trace, Kingfort, DE 70686           US
1    07681 Ian Heights Apt. 855, Camachohaven, AL 2...           US
2    23622 Peter Camp Suite 968, Melanieview, TN 62387           US
3    516 John Vista Suite 215, Lake Darrellchester,...           US
4            0335 Tiffany Via, Patriciaburgh, IL 92620           US
..                                                 ...          ...
995  51547 Graham Crescent Apt. 708, Smithfort, ND ...           US
996                   PSC 4009, Box 3808, APO AP 26991  US military
997        32160 Nicholas Rapid, East Joseph, ND 31097           US
998  76252 Nicole Corners Apt. 116, Elliottport, NC...           US
999  0116 Wagner Creek Suite 786, New Adrian, CT 44160           US

[1000 rows x 2 columns]

## 🔍 Post Address Cleaning Validation:

In [12]:

query = '''
        select address , street, state_code, state , country,zipcode
        # select * 
        from aligned_customers_ws 
        where street is null
        OR state_code is null 
        OR state is null
        OR country is null
        ;
        '''
add = pd.read_sql(query,engine)
add

Empty DataFrame
Columns: [address, street, state_code, state, country, zipcode]
Index: []

All address components were successfully parsed and populated:
- No NULLs found in `street`, `state_code`, `state`, or `country` columns
- Military and standard US addresses were handled with regex 

## 5.💸Cleaning `Branches` Table:

#### Phone_Number Column:

In [ ]:
from sqlalchemy import text

with engine.begin() as conn:
    conn.execute(text(
        '''
        Alter table branches add column cleaned_contact_number VARCHAR (80);
        '''
    ))

In [13]:
# Updating the cleaned_contact_number post cleaning task:

from sqlalchemy import text

with engine.begin() as conn:
    conn.execute(text(
        '''
        # select contact_number,regexp_replace(regexp_replace(contact_number,'[^+0-9x]',''),'x',' x')  from branches
        # where contact_number like '+1%'
        # ; 
        
        
        update branches
        set cleaned_contact_number = regexp_replace(regexp_replace(contact_number,'[^+0-9x]',''),'x',' x')
        where cleaned_contact_number is null
        ;
       '''
))
query = '''
        # select COUNT(contact_number), COUNT(cleaned_contact_number)
        select contact_number,cleaned_contact_number
        from branches where cleaned_contact_number is not null
        ; 
        '''
branch = pd.read_sql(query,engine)
branch

contact_number cleaned_contact_number
0  +1-523-918-8676x22735    +15239188676 x22735
1           340.620.1663             3406201663
2           401.421.5634             4014215634
3   001-526-362-8862x335     0015263628862 x335
4       044.455.1434x421        0444551434 x421
5           175-494-4957             1754944957
6   +1-421-884-2463x8151     +14218842463 x8151
7          (911)647-9246             9116479246
8      299-872-4735x3998       2998724735 x3998
9      591.530.0751x4518       5915300751 x4518

## 🔍Post Branch-Phone Cleaning Validation:

- missing_cleaned_contacts = 0/10
 → All branch contact numbers were successfully cleaned.

## 📟Accounts Table Validation:

In [14]:
# Customers with no accounts:

query = '''
            SELECT c.customer_id,c.cleaned_name
            FROM aligned_customers_ws AS c
            LEFT JOIN aligned_account_ws AS a ON c.customer_id = a.customer_id
            WHERE a.account_id IS NULL;
        '''
no_accounts = pd.read_sql(query,engine)
no_accounts

customer_id       cleaned_name
0    CUST000289         Lisa Allen
1    CUST000290      Briana Murray
2    CUST000291        David Smith
3    CUST000292     Stephanie Byrd
4    CUST000293        Jeremy Reed
..          ...                ...
707  CUST000996      Brittany Ward
708  CUST000997     Edward Stanley
709  CUST000998  Christina Johnson
710  CUST000999       Edgar Miller
711  CUST001000  Elizabeth Sanders

[712 rows x 2 columns]

## ⍰Customers with no financial footprint:
- This could be due to:

🆕 Newly onboarded customers whose accounts haven't been created yet.

💤 Inactive or dormant customers whose accounts were closed or are not linked properly.

## Feature Engineering : Accounts Table 🧮

### 💱Loan-Balance Ratio :
It measures riskiness of a customer based on debt (loan) versus available funds (balance).

In [ ]:
#Adding Loan-Balance Ratio column to the accounts table:

with engine.begin() as conn:
    conn.execute(text(
                    '''
                    alter table aligned_account_ws add column loan_balance_ratio varchar(60);
                    '''
    ))

In [36]:
#Updating the values of loan_balance_ratio

# with engine.begin() as conn:
#     conn.execute(text(
#         '''
#         update aligned_account_ws 
#         set loan_balance_ratio = CASE WHEN loan_amount = 0 THEN null ELSE ROUND(loan_amount / balance, 2) END 
#         where loan_balance_ratio is null 
#         ;
#         '''
#     ))

query = '''
        select a.account_id,c.cleaned_name,a.balance,a.loan_amount,a.loan_balance_ratio
        from aligned_account_ws as a
        join aligned_customers_ws as c on a.customer_id = c.customer_id
        where loan_balance_ratio is not null;
        '''
loan_ratio = pd.read_sql(query,engine)
loan_ratio

account_id      cleaned_name       balance  loan_amount  \
0    CUR000000001      Allison Hill  369877.52670        73933   
1    LOA000000003   Angie Henderson  161796.05100       877843   
2    LOA000000116     Brian Ramirez  107642.46490       751786   
3    SAV000000009  Margaret Hawkins  232629.25730       285964   
4     FD000000007       Ethan Adams  157512.57810       103367   
..            ...               ...           ...          ...   
100  SAV000000094    Shelia Wallace  194723.22200       125729   
101   FD000000078     Elaine Brooks  100181.93830       500450   
102  CUR000000092     Elaine Brooks  100181.93830       500450   
103  CUR000000093     Joseph Knight   73873.03372       780608   
104  LOA000000127    Carolyn Miller  218497.77300        69958   

    loan_balance_ratio  
0                  0.2  
1                 5.43  
2                 6.98  
3                 1.23  
4                 0.66  
..                 ...  
100               0.65  
101                  5  
102                  5  
103              10.57  
104               0.32  

[105 rows x 5 columns]

### ⚠️ Loan Risk Label:
This typcially helps to assess how safe or risky a customer,loan or investment is.

In [ ]:
#Altering table to add another column of "loan risk" :

with engine.begin() as conn:
    conn.execute(text(
                    '''
                        # alter table aligned_account_ws add column loan_risk_label varchar(80);
                    '''
        
    ))

query = '''
            select * from aligned_account_ws;
        '''
risk = pd.read_sql(query,engine)
risk

In [37]:
# with engine.begin() as conn:
#     result = conn.execute(text(
#                     '''
#                     update aligned_account_ws
#                     set loan_risk_label = case when loan_balance_ratio  = 0 then '✅ Low_risk'
#                     when loan_balance_ratio <=0.5 then '🟢 Safe_risk'
#                     when loan_balance_ratio <=1.0 then '🟡 Moderate risk'
#                     when loan_balance_ratio <=5.0 then '🔴 High risk'
#                     when loan_balance_ratio >5.0 then '🚨 Very high risk'
#                     end 
#                     where loan_risk_label is not null
#                     ;
#                     '''
#             ))
# result 
query = '''
    SELECT loan_amount, balance, loan_balance_ratio,loan_risk_label
    FROM aligned_account_ws 
    WHERE loan_balance_ratio IS NOT NULL;
'''
risk = pd.read_sql(query, engine)
risk

loan_amount      balance loan_balance_ratio   loan_risk_label
0          73933  369877.5267                0.2       🟢 Safe_risk
1         389041  262352.7733               1.48       🔴 High risk
2         404466  340645.2881               1.19       🔴 High risk
3         244225   34167.1000               7.15  🚨 Very high risk
4         338121  198459.4272                1.7       🔴 High risk
..           ...          ...                ...               ...
100       517908   76195.7400                6.8  🚨 Very high risk
101        96260  283358.1011               0.34       🟢 Safe_risk
102       317822   33948.9900               9.36  🚨 Very high risk
103       344793  348710.8653               0.99   🟡 Moderate risk
104       751031  176023.5064               4.27       🔴 High risk

[105 rows x 4 columns]

## Validating the featured column of Loan_Balance Ratio:


❌ Loan ratio has 399 NULLs — investigation shows these are customers with no loan amount. Will exclude them from    risk ratio calculation by design.

✅ 105 customers have valid loan_balance_ratio values where both loan_amount > 0. These entries are correctly included for credit risk profiling.

##  💳 Credit Card Simulation: 

In [ ]:
# Compute loan_to_balance_ratio:

# loan_balance_ratio = round((loan_amount / balance),2)

# Feature engineering for synthetic cibil_score:

with engine.begin() as conn:
    conn.execute(text(
        '''
    create table Cibil_accounts as     
    (
        select  customer_id,
        CASE 
            WHEN t1.loan_amount = 0 THEN 30
            WHEN t1.loan_amount / (t1.balance + 1) <= 0.5 THEN 25
            WHEN t1.loan_amount / (t1.balance + 1) <= 1.0 THEN 15
            WHEN t1.loan_amount / (t1.balance + 1) <= 5.0 THEN 10
            ELSE 5
          END AS loan_score,
            -- 💳 Score by credit card usage
          CASE 
            WHEN t1.credit_card = 'Yes' THEN 25
            ELSE 10
          END AS credit_score,
            -- 💰 Score by balance buckets
          CASE 
            WHEN t1.balance >= 100000 THEN 30
            WHEN t1.balance >= 50000 THEN 20
            WHEN t1.balance >= 10000 THEN 15
            ELSE 10
          END AS balance_score,
         -- 🧠 Final Synthetic CIBIL Score (sum of above)
          (
            CASE 
              WHEN t1.loan_amount = 0 THEN 30
              WHEN t1.loan_amount / (t1.balance + 1) <= 0.5 THEN 25
              WHEN t1.loan_amount / (t1.balance + 1) <= 1.0 THEN 15
              WHEN t1.loan_amount / (t1.balance + 1) <= 5.0 THEN 10
              ELSE 5
            END
            +
            CASE 
              WHEN t1.credit_card = 'Yes' THEN 25
              ELSE 10
            END
            +
            CASE 
              WHEN t1.balance >= 100000 THEN 30
              WHEN t1.balance >= 50000 THEN 20
              WHEN t1.balance >= 10000 THEN 15
              ELSE 10
            END
          ) AS cibil_score
    FROM aligned_account_ws AS t1
    where account_type = 'Loan'
    )
    ;
'''
    ))

query = '''
        select * from cibil_accounts ;
        '''
cibil = pd.read_sql(query,engine)
cibil

### 🧮 CIBIL Score Simulation for Loan Accounts:

This query creates a new table cibil_accounts to estimate a synthetic CIBIL score for customers holding loan accounts, based on:

🔍 Key Components:
1. Loan Score (loan_score)
- Based on how much loan a customer has compared to their balance:

- Lower loan-to-balance ratio → Higher score

- No loan → Full score (30)

2. Credit Score (credit_score)
- Based on credit card availability:

- Has credit card → 25 points

- No credit card → 10 points

3. Balance Score (balance_score)
- Higher balance → Higher score:

- ₹1,00,000+ → 30 points

- ₹50,000+ → 20

- ₹10,000+ → 15

- Else → 10

4. Final cibil_score
- Sum of the 3 components above.

## CIBIL Score Ratings:
To enrich customer credit profiling, the dataset cibil_accounts was updated to assign a credit_rating based on their cibil_score using defined score bands:

In [ ]:
# ALtering table addition to credit_ratings and tagging a constraint of Primary Key

with engine.begin() as conn:
    conn.execute(text(
                    '''
                        # alter table cibil_accounts add column credit_ratings varchar(80);
                        #alter table cibil_accounts add constraint pk_cibil primary key (customer_id); 
                    '''
        
    ))

In [40]:
with engine.begin() as conn:
    conn.execute(text(
                    '''
                    # update cibil_accounts 
                    # set credit_rating = case when cibil_score between 80 and 100 then 'Excellent Rating'
                    # when cibil_score between 65 and 79 then 'Good'
                    # when cibil_score between 50 and 64 then 'Fair'
                    # when cibil_score between 35 and 49 then 'Poor'
                    # when cibil_score < 35 then 'Risky/Bad' end 
                    # where credit_rating is null ;
                    '''
    ))

query = '''
            select * from cibil_accounts;
        '''
score = pd.read_sql(query,engine)
score

customer_id  loan_score  credit_score  balance_score  cibil_score  \
0    CUST000001          30            10             30          700   
1    CUST000002          30            10             30          700   
2    CUST000003           5            25             30          600   
3    CUST000006          30            10             30          700   
4    CUST000007          30            10             30          700   
..          ...         ...           ...            ...          ...   
126  CUST000279          10            10             30          500   
127  CUST000283          30            10             30          700   
128  CUST000285          30            10             30          700   
129  CUST000286          30            25             30          850   
130  CUST000288          25            10             30          650   

        credit_rating  
0                Good  
1                Good  
2                Fair  
3                Good  
4                Good  
..                ...  
126              Fair  
127              Good  
128              Good  
129  Excellent Rating  
130              Good  

[131 rows x 6 columns]

### ✅ Data readiness confirmed

All customer, account, and transaction data has been cleaned and enhanced with new features like domain status, phone normalization, and credit risk metrics. 

This dataset is now ready for SQL-based EDA and visual analytics.

### ➕ Check out the companion notebook for Bank_EDA_Main.ipynb ➡️ culminating to:
- segmentation
- aggregation
- dashboarding.

👉 **Proceed to Main Analysis Notebook:**  
📘 [Explore the Full EDA on Bank Customers →](https://github.com/nikita94-stack/bank_eda_project/blob/main/Bank_Customers.ipynb)